From fc5b04bb609323131e55a451618c3b4d51443a18 Mon Sep 17 00:00:00 2001
From: Amelie Delaunay <amelie.delaunay@foss.st.com>
Date: Thu, 23 Mar 2023 15:52:40 +0100
Subject: [PATCH 0102/1141] dmaengine: stm32-dma3: prevent LL refactoring
 thanks to DT configuration

stm32-dma3 driver refactors the linked-list in order to address the memory
with the highest possible data width.
It means that it can introduce up to 2 linked-list items. One with a
transfer length multiple of channel maximum burst length and so with the
highest possible data width. And an extra one with the latest bytes, with
lower data width.
Some devices (e.g. FMC ECC) don't support having several transfers instead
of only one.
So add the possibility to prevent linked-list refactoring, when bit 17 of
the 'DMA transfer requirements' bit mask is set in device tree.

Signed-off-by: Amelie Delaunay <amelie.delaunay@foss.st.com>
Change-Id: I865c4da031d3712ea08f52bfb1e1343634b72ba5
---
 drivers/dma/stm32-dma3.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/drivers/dma/stm32-dma3.c
+++ b/drivers/dma/stm32-dma3.c
@@ -300,6 +300,7 @@ enum g_data_width {
 #define STM32_DMA3_DT_PFREQ		BIT(9)		/* CTR2_PFREQ */
 #define STM32_DMA3_DT_TCEM		GENMASK(13, 12)	/* CTR2_TCEM */
 #define STM32_DMA3_DT_NOPACK		BIT(16)		/* CTR1_PAM */
+#define STM32_DMA3_DT_NOREFACT		BIT(17)
 
 /* .tr_conf_ext */
 #define STM32_DMA3_DT_SAO		GENMASK(12, 0)	/* CTR3_SAO */
@@ -1425,7 +1426,8 @@ static struct dma_async_tx_descriptor *s
 	size_t next_size, offset;
 	u32 max_block_size = STM32_DMA3_MAX_BLOCK_SIZE;
 	u32 count, i, ccr, ctr1, ctr2, ctr3;
-	bool prevent_refactor = !!FIELD_GET(STM32_DMA3_DT_NOPACK, chan->dt_config.tr_conf);
+	bool prevent_refactor = !!FIELD_GET(STM32_DMA3_DT_NOPACK, chan->dt_config.tr_conf) ||
+				!!FIELD_GET(STM32_DMA3_DT_NOREFACT, chan->dt_config.tr_conf);
 
 	/* TODO: Check if channel is busy ? */
 
@@ -1504,7 +1506,8 @@ static struct dma_async_tx_descriptor *s
 	dma_addr_t sg_addr, dev_addr, src, dst;
 	u32 max_block_size = STM32_DMA3_MAX_BLOCK_SIZE;
 	u32 i, j, count, ccr, ctr1, ctr2, ctr3;
-	bool prevent_refactor = !!FIELD_GET(STM32_DMA3_DT_NOPACK, chan->dt_config.tr_conf);
+	bool prevent_refactor = !!FIELD_GET(STM32_DMA3_DT_NOPACK, chan->dt_config.tr_conf) ||
+				!!FIELD_GET(STM32_DMA3_DT_NOREFACT, chan->dt_config.tr_conf);
 	int ret;
 
 	/* TODO: Check if channel is busy ? */
